### ISSUE TITLE
JSON library question

### ISSUE THREAD

**Issue by sptrakesh** [matt-42/lithium#21]
Hello,

I am trying out the JSON library to use with a Grafana JSON datasource project I am starting.  I set up some simple structs to model the data as described at https://grafana.com/grafana/plugins/grafana-simple-json-datasource and started a simple unit test for parsing.

With the first *Query API* sample payload, I am getting only partial results.  I get the `panelId` and `range` parts, but nothing after.  Wondering what I am missing.

```
GIVEN("Sample query from grafana documentation")
  {
    const auto json = R"({
  "panelId": 1,
  "range": {
    "from": "2016-10-31T06:33:44.866Z",
    "to": "2016-10-31T12:33:44.866Z",
    "raw": {
      "from": "now-6h",
      "to": "now"
    }
  },
  "rangeRaw": {
    "from": "now-6h",
    "to": "now"
  },
  "interval": "30s",
  "intervalMs": 30000,
  "targets": [
     { "target": "upper_50", "refId": "A", "type": "timeserie" },
     { "target": "upper_75", "refId": "B", "type": "timeserie" }
  ],
  "adhocFilters": [
    "key": "City"
    "operator": "=",
    "value": "Berlin"
  ]
  "format": "json",
  "maxDataPoints": 550
})";

    WHEN("Parsing query")
    {
      auto query = spt::model::Query{};
      auto obj = li::json_object( s::panelId,
          s::range = li::json_object( s::from, s::to, s::raw = li::json_object( s::from, s::to ) ),
          s::rangeRaw = li::json_object( s::from, s::to ),
          s::interval, s::intervalMs,
          s::targets = li::json_vector( s::target, s::refId, s::type ),
          s::adhocFilters = li::json_vector( s::key, s::value ),
          s::format, s::maxDataPoints );
      obj.decode( json, query );

      THEN("Query structure populated")
      {
        REQUIRE(query.panelId == 1);
        REQUIRE(query.range.from == "2016-10-31T06:33:44.866Z");
        REQUIRE(query.range.to == "2016-10-31T12:33:44.866Z");
        REQUIRE(query.targets.size() == 2);
        REQUIRE(query.interval == "30s");
        REQUIRE(query.format == "json");
        REQUIRE(query.maxDataPoints == 550);
      }

      THEN("Serialise structure back to JSON")
      {
        std::cout << obj.encode( query ) << '\n';
      }
    }
```

```
{"panelId":1,"range":{"from":"2016-10-31T06:33:44.866Z","to":"2016-10-31T12:33:44.866Z","raw":{"from":"now-6h","to":"now"}},"rangeRaw":{"from":"","to":""},"interval":"","intervalMs":0,"targets":[],"adhocFilters":[],"format":"","maxDataPoints":0}
```

On a related note, is it possible to use `std::optional<custom struct>`?  When I tried to make `rangeRaw` an optional,  I get a compiler error about missing `error`:

```
lithium_json.h:626:58: error: no member named 'error' in 'li::CANNOT_FIND_REQUESTED_MEMBER_IN_TYPE<s::from_t, std::__1::optional<spt::model::RangeRaw> &>'
      return CANNOT_FIND_REQUESTED_MEMBER_IN_TYPE<S, T>::error;
```

**Comment by matt-42** [comment]
Trying to reproduce the bug.

**Comment by sptrakesh** [comment]
Thanks for taking a look.  In case this is useful, here are my initial structures around which I am building unit tests.

```
LI_SYMBOL(from)
LI_SYMBOL(to)
LI_SYMBOL(datasource)
LI_SYMBOL(iconColor)
LI_SYMBOL(enable)
LI_SYMBOL(showLine)
LI_SYMBOL(query)
LI_SYMBOL(range)
LI_SYMBOL(annotation)
LI_SYMBOL(key)
LI_SYMBOL(value)
LI_SYMBOL(time)
LI_SYMBOL(title)
LI_SYMBOL(tags)
LI_SYMBOL(text)
LI_SYMBOL(target)
LI_SYMBOL(adhocFilters)
LI_SYMBOL(refId)
LI_SYMBOL(panelId)
LI_SYMBOL(interval)
LI_SYMBOL(intervalMs)
LI_SYMBOL(targets);
LI_SYMBOL(format);
LI_SYMBOL(maxDataPoints);
LI_SYMBOL(coordinates);
LI_SYMBOL(timestamp);
LI_SYMBOL(metadata);
LI_SYMBOL(columns);
LI_SYMBOL(rows);
LI_SYMBOL(code);
LI_SYMBOL(cause);

LI_SYMBOL(raw);
LI_SYMBOL(rangeRaw);

namespace spt::model
{
  struct RangeRaw
  {
    std::string from;
    std::string to;
  };

  struct Range
  {
    std::string from;
    std::string to;
    RangeRaw raw;
  };

  struct Annotation
  {
    std::string name;
    std::string datasource;
    std::string iconColor;
    bool enable;
    bool showLine;
    std::string query;
  };

  struct AnnotationsReq
  {
    Range range;
    Annotation annotation;
  };

  struct Tag
  {
    std::string key;
    std::string value;
  };

  struct AnnotationResponse
  {
    // The original annotation sent from Grafana.
    Annotation annotation;
    // Time since UNIX Epoch in milliseconds. (required)
    int64_t time;
    // The title for the annotation tooltip. (required)
    std::string title;
    // Tags for the annotation. (optional)
    std::vector<Tag> tags;
    // Text for the annotation. (optional)
    std::string text;
  };

  struct Target
  {
    std::string target;
    std::string refId;
    std::string type;
  };

  struct Query
  {
    int panelId;
    Range range;
    RangeRaw rangeRaw;
    std::string interval;
    int intervalMs;
    std::vector<Target> targets;
    std::vector<Tag> adhocFilters;
    std::string format;
    int maxDataPoints;
  };

  struct LocationValue
  {
    std::string type;
    std::vector<double> coordinates;
  };

  struct Timestamp
  {
    std::string type;
    std::string value;
  };

  struct Metadata
  {
    Timestamp timestamp;
  };

  struct Column
  {
    std::string text;
    std::string type;
  };

  struct Row
  {
    std::string type;
    LocationValue value;
    Metadata metadata;
  };

  struct LocationResponse
  {
    std::vector<Column> columns;
    std::vector<std::vector<Row>> rows;
    std::string type;
  };

  struct TagKey
  {
    std::string text;
    std::string type;
  };

  struct TagValue
  {
    std::string key;
  };

  struct Error
  {
    // Usually a HTTP status code associated with the error.
    int code;

    // A context specific error string describing the cause of the error.
    std::string cause;
  };
}
````

**Comment by matt-42** [comment]
I found problem in lithium and in your code. I'm preparing a fix for both

**Comment by matt-42** [comment]
Fixed in 8a4cdeafde2960aaf097c5fa01bd30d1a07a8eab . I fixed your code here :
https://github.com/matt-42/lithium/blob/8a4cdeafde2960aaf097c5fa01bd30d1a07a8eab/libraries/json/tests/graphana.cc

Don't forget to check for errors, they can help you to debug :

```
auto err = obj.decode(json, query);
if (err) std::cout << err.what << std::endl;
```
